### PR TITLE
hicubenorm tests added

### PR DESCRIPTION
**Pull request by Kelvinrr**
## Description

Added GTests for hicube norm

## Motivation and Context

983Mb

## How Has This Been Tested?

All new tests pass locally. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have read and agree to abide by the [Code of Conduct](https://usgs-astrogeology.github.io/code/)
- [ ] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/mro/apps/hicubenorm/hicubenorm.cpp [hunk 1]
```diff
@@ -0,0 +1,760 @@
...
+    }
+    out.close();
+    }
+
+    //********************************************************
+    // Gather statistics from a PVL input file
+    //*******************************************************
+    void PVLIn(const Isis::FileName &filename) {
+    Pvl pvlFileIn;
+    pvlFileIn.read(filename.expanded());
```
Comment (Kelvinrr) [inline isis/src/mro/apps/hicubenorm/hicubenorm.cpp:326]: this was originally filename.name(), which seemed like a bug, so I replaced it with expanded.

File: isis/tests/FunctionalTestsHiCubeNorm.cpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+  Cube oCube(outCubeFileName, "r"); 
+  oCube.reopen("rw");
+
+  Histogram *oCubeStats = oCube.histogram();
+  
+  ASSERT_DOUBLE_EQ(oCubeStats->Average(), 602.50000120516233); 
+  ASSERT_DOUBLE_EQ(oCubeStats->Sum(), 766032961.53227234); 
+  ASSERT_EQ(oCubeStats->ValidPixels(), 1271424);  
+  ASSERT_DOUBLE_EQ(oCubeStats->StandardDeviation(), 347.55180258334673);
```
Comment (Kelvinrr) [inline isis/tests/FunctionalTestsHiCubeNorm.cpp:169]: Wasn't sure how to best test DNs, using stats seems like a decent answer?

**Review by AustinSanders — DISMISSED** [review]
Looks good

File: isis/src/mro/apps/hicubenorm/hicubenorm.cpp (hunk 1 above)
Comment (scsides) [reply isis/src/mro/apps/hicubenorm/hicubenorm.cpp:326]: Agreed

File: isis/src/mro/apps/hicubenorm/hicubenorm.h [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+#include "Cube.h"
+#include "UserInterface.h"
+
+namespace Isis {
+  extern void hicubenorm(Cube *cube, UserInterface &ui);
+  extern void hicubenorm(UserInterface &ui);
+}
+
+#endif
```
Comment (scsides) [inline isis/src/mro/apps/hicubenorm/hicubenorm.h]: github is upset there isn't a newline at the end of the file. I know the new C++ doesn't require it, but it was undefined behavior prior to c++ 11 (I think), and it causes odd problems with using tools to parse files that are looking for newlines and doing something different for EOFs.

This is true of other files too.

File: isis/src/mro/apps/hicubenorm/hicubenorm.cpp [hunk 4]
```diff
@@ -0,0 +1,760 @@
...
+        else {
+        median = dFilter;
+        }
+        dInFilter.clear();
+        dOrigFilter.clear();
+        dTempFilter.clear();
+        dFilter.clear();
+    }
+    }
+}
```
Comment (scsides) [inline isis/src/mro/apps/hicubenorm/hicubenorm.cpp]: This and some of the other files do not have a NL at the EOF. Github is upset about this.

Although this is not required after C++11 (I think), it is still nice to have for other tools that may behave different when there isn't one.

File: isis/tests/Fixtures.cpp [hunk 5]
```diff
@@ -258,4 +258,139 @@ namespace Isis {
...
+        for(int i = 0; i < line.size(); i++) {
+          line[i] = (double)(i+1);
+        }
+        testCube->write(line);
+    }
+    testCube->reopen("rw");
+ 
+    // need to remove old camera pointer 
+    delete testCube;
+    // This is now a Hayabusa cube
```
Comment (scsides) [inline isis/tests/Fixtures.cpp]: This is now an MRO HiRISE cube

File: isis/tests/FunctionalTestsHiCubeNorm.cpp [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  ASSERT_DOUBLE_EQ((double)res[319], 40);
+
+  ASSERT_DOUBLE_EQ((double)res[9627], 5.1018076194189);
+  ASSERT_DOUBLE_EQ((double)res[9628], 5.1018076194189);
+  ASSERT_DOUBLE_EQ((double)res[9629], 0);
+  ASSERT_DOUBLE_EQ((double)res[9630], 1204);
+  ASSERT_DOUBLE_EQ((double)res[9631], 1204);
+
+  Cube oCube(outCubeFileName, "r"); 
+  oCube.reopen("rw");
```
Comment (scsides) [inline isis/tests/FunctionalTestsHiCubeNorm.cpp]: What am I missing? Why not just open it 'rw' to begin with?

File: isis/tests/FunctionalTestsHiCubeNorm.cpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  UserInterface options(APP_XML, args);
+  try {
+    hicubenorm(testCube, options);
+  }
+  catch (IException &e) {
+    FAIL() << "Unable to process HRISE image: " << e.what() << std::endl;
+  }
+
+  Cube oCube(outCubeFileName, "r"); 
+  oCube.reopen("rw");
```
Comment (scsides) [inline isis/tests/FunctionalTestsHiCubeNorm.cpp]: Why does this need to be 'rw' ?

File: isis/tests/FunctionalTestsHiCubeNorm.cpp (hunk 2 above)
Comment (scsides) [reply isis/tests/FunctionalTestsHiCubeNorm.cpp:169]: Without testing individual DNs, stats should be a good compromise.

**Review by scsides — COMMENTED** [review]
Mostly just questions. Happy to approve once I understand.

File: isis/tests/FunctionalTestsHiCubeNorm.cpp (hunk 6 above)
Comment (Kelvinrr) [reply isis/tests/FunctionalTestsHiCubeNorm.cpp]: I'll go back and double check which reopens are required, but while writing this I needed to basically reopen after modifying the cube or else I end up with a segfault. I removed some of the cube changes so I'll which can go as I may have left the reopens in.

File: isis/tests/FunctionalTestsHiCubeNorm.cpp (hunk 7 above)
Comment (Kelvinrr) [reply isis/tests/FunctionalTestsHiCubeNorm.cpp]: probably a left over from another version of the test, set to read only.

**Review by scsides — APPROVED** [review]

**Pull request merged**